src/main/java/io/confluent/connect/hdfs/wal/FSWAL.java

```diff
@@ -183,11 +188,156 @@ private void commitEntriesToStorage(Map<WALEntry, WALEntry> entries) {
+      if (storage.exists(logFile)) {
+        log.trace("Restoring offset from WAL file: {}", logFile);
+        if (reader == null) {
+          reader = newWalFileReader(logFile);
+        } else {
+          // reset read position after apply()
+          reader.seekToFirstRecord();
+        }
+        List<String> committedFileBatch = getLastFilledBlockFromWAL(reader);
+        // At this point the committedFilenames list will contain the
+        // filenames in the last BEGIN-END block of the file. Find the latest offsets among these.
+        latestOffset = getLatestOffsetFromList(committedFileBatch);
+      }
+
+      // attempt to use old log file if recent WAL is empty or non-existent
+      if (latestOffset == null && storage.exists(oldWALFile)) {
+        log.trace("Could not find offset in log file {}. Using {} instead", logFile, oldWALFile);
+        try (Reader oldFileReader = newWalFileReader(oldWALFile)) {
+          List<String> committedFileBatch = getLastFilledBlockFromWAL(oldFileReader);
+          latestOffset = getLatestOffsetFromList(committedFileBatch);
+        }
+      }
+      return latestOffset;
+    } catch (IOException e) {
+      log.warn(
+          "Error restoring offsets from either {} or {} WAL files: {}",
+          logFile,
+          oldWALFile,
+          e.getMessage()
+      );
+      return null;
+    }
+  }
+
+  /**
+   * Extract the last filled BEGIN-END block of entries from the WAL.
+   *
+   * @param reader the WAL file reader
+   * @return the last batch of entries, may be empty if the WAL
+   *         is empty or only contains empty BEGIN-END blocks
+   * @throws IOException error on reading the WAL file
+   */
+  private List<String> getLastFilledBlockFromWAL(Reader reader) throws IOException {
+    // In a WAL entry the temp filenames (keys) don't contain offset info,
+    // so we only need to track the committed filename (values).
+    List<String> committedFilenames = new ArrayList<>();
+    // enables skipping corrupted blocks that are missing an END marker,
+    // only valid blocks will be moved to committedFilenames
+    List<String> tempFilenames = new ArrayList<>();
+    WALEntry key = new WALEntry();
+    WALEntry value = new WALEntry();
+
+    // whether a BEGIN-END block was started
+    boolean entryBlockStarted = false;
+
+    // The entry with the latest offsets will be in the last BEGIN-END block of the file.
+    // There may be empty BEGIN and END blocks as well, skip over these and only keep the
+    // entries of the last filled block.
+    while (reader.next(key, value)) {
+      String keyName = key.getName();
+      if (keyName.equals(beginMarker)) {
+        tempFilenames.clear();
+        entryBlockStarted = true;
+      } else if (keyName.equals(endMarker)) {
+        if (entryBlockStarted && !tempFilenames.isEmpty()) {
+          // only save non-empty blocks
+          committedFilenames.clear();
+          committedFilenames.addAll(tempFilenames);
```

Comment:
I had synced with @erdody on this previously, the idea of `clear()` and `addAll()` here is to minimize the memory allocation. The effect is equivalent.

Reply:
The benefits are not as clear cut. For such optimizations it's worth checking the underlying implementations. 
* `addAll` creates a copy of the source array internally and assigns it to a local object. This local array will also be GC'ed once this method exits because it's copied to the destination array, it's not replacing it, because the method adds all the source elements to an existing array that potentially has more elements (not the case here). 
* `ArrayList(Collection ...)` creates such a copy but normally it assigns this to the destination array (the one that is constructed). So what is GC'ed is the old array in your code (as you've correctly observed). 

Therefore, besides a very vague and debatable advantage related to cache locality (which here might be even smaller because this is not a member variable array but still a local array within a method) we can't really be sure which approach works best. The number of arrays being GC'ed in the end is the same. And all that, without even saying that this is heavily dependent on whether the new arrays fit within the old one or the old one needs to get expanded more often than not. 

Given all that, unless this was surfaced as a bottleneck during our testing (which could include microbenchmarks with JMH if we wanted a quick assessment) or regular operation of the connector, I'd choose the most simple way. To me both work, so up to you. (As a last note, also worth saying that in modern jvm's allocation of new arrays has improved).